File: internal/wal/checksum_test.go
package wal

import (
	"os"
	"path/filepath"
	"testing"

	"bigtable/internal/record"
)

func TestWALChecksumDetectsCorruption(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "wal.log")

	w, err := Open(path)
	if err != nil {
		t.Fatalf("open: %v", err)
	}

	if err := w.Append(record.NewPut([]byte("k1"), []byte("v1"), 1)); err != nil {
		t.Fatalf("append: %v", err)
	}
	if err := w.Close(); err != nil {
		t.Fatalf("close: %v", err)
	}

	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read file: %v", err)
	}

	if len(data) < 12 {
		t.Fatal("wal too small for corruption test")
	}

	data[10] ^= 0xFF

	if err := os.WriteFile(path, data, 0o644); err != nil {
		t.Fatalf("rewrite corrupt wal: %v", err)
	}

	w2, err := Open(path)
	if err != nil {
		t.Fatalf("reopen: %v", err)
	}
	defer w2.Close()

	err = w2.ReplayAll(func(r record.Record) error { return nil })
	if err == nil {
		t.Fatal("expected checksum error, got nil")
	}
}

File: internal/wal/checksum_test.cpp
#include "wal.h"
#include <gtest/gtest.h>
#include <filesystem>
#include <fstream>
#include "../record/record.h"

using namespace bigdb::wal;
using namespace bigdb::record;
namespace fs = std::filesystem;

class WALChecksumTest : public ::testing::Test {
protected:
    void SetUp() override {
        temp_dir = testing::TempDir() + "/wal_checksum_test_" + std::to_string(std::chrono::system_clock::now().time_since_epoch().count());
        fs::create_directories(temp_dir);
    }
    void TearDown() override {
        fs::remove_all(temp_dir);
    }
    std::string temp_dir;
};

TEST_F(WALChecksumTest, ChecksumDetectsCorruption) {
    std::string path = temp_dir + "/wal.log";
    auto w = WAL::Open(path);
    ASSERT_NE(w, nullptr);

    EXPECT_NO_THROW(w->Append(Record::new_put({'k', '1'}, {'v', '1'}, 1)));
    w->Close();

    std::fstream f(path, std::ios::in | std::ios::out | std::ios::binary);
    ASSERT_TRUE(f.good());
    
    f.seekg(0, std::ios::end);
    size_t size = f.tellg();
    ASSERT_GE(size, 12);

    f.seekg(10, std::ios::beg);
    char byte;
    f.read(&byte, 1);
    
    byte ^= 0xFF;
    
    f.seekp(10, std::ios::beg);
    f.write(&byte, 1);
    f.close();

    auto w2 = WAL::Open(path);
    ASSERT_NE(w2, nullptr);

    EXPECT_THROW(w2->ReplayAll([](const Record&) {}), std::runtime_error);
}
